### PR TITLE
OpenCL has same initial case seeding logic

### PR DESCRIPTION
**Pull request by rory-improbable**
This PR changes the way the seeding of initial cases works in the OpenCL model, so that instead of adding the cases to the initial snapshot for the first 10 days all at once, then starting to run the model at timestep 11, there is now a new function on the `Simulator` class called `step_with_seeding()`, which is run for the first 10 timesteps instead of the normal `step()` function.

this `step_with_seeding()` function works by selecting a number of people in high risk msoas based on the initial cases data (by reading from the initial cases csv file), then setting the "hazard" value for these people to be a very high value (the maximum float32 value), so that this gets mapped to an infection probability of 1.0. This way we can then just run the individual `people_update_statuses` OpenCL kernel and re-use the exact same logic that the model normally uses, but only running this one kernel by itself.

# NB:
Once this PR is merged you will have to update any jupyter notebooks you're using to remove the line which calls the old seeding function from the snapshot object, since now the seeding is done by the Simulator class, eg. remove this line
```
snapshot.seed_initial_infections(num_seed_days=num_seed_days, data_dir=data_dir)
```
Additionally the Simulator class now just takes a path to the OpenCL directory, not the kernel directory. 
I've already made these changes in the experiments/analyse_opencl_results.ipynb notebook, so you can just copy the code from there.

## Results
I haven't looked at the results numbers in much detail, so it's definitely worth someone else having a look to check that they're sensible. However this should mean the results are closer to the original Python/R model. 

### Results with previous GAM initial cases file:
image

### Results with new GAM initial cases file:
image

### CODE REVIEW

**Comment by Sparrow0hawk** [comment]
@nickmalleson or @spoonerf are either of you OK to do some validation with the python model versus these changes?

**Review by Sparrow0hawk — COMMENTED** [review]
I'll have a quick check of the notebook later but all code looks good!

File: microsim/opencl/ramp/snapshot.py [hunk 1]
```diff
@@ -35,7 +32,7 @@ def zeros(cls, nplaces, npeople, nslots):
         npeople = np.uint32(npeople)
         nslots = np.uint32(nslots)
         time = np.uint32(0)
-        area_codes = np.full("", npeople)
+        area_codes = np.full(npeople, "E02004129")
```
[the remark this answers is not in the document]
Comment (rory-improbable) [reply microsim/opencl/ramp/snapshot.py:35]: yep this is a hardcoded Devon postcode, this function is only really used to create an empty snapshot for testing though, so hopefully it shouldn't matter too much what the value is.

[the remark this answers is not in the document]
Comment (Sparrow0hawk) [reply microsim/opencl/ramp/snapshot.py:35]: Got it, indeed shouldn't matter much.

**Review by spoonerf — APPROVED** [review]
This looks good to me! 

A bit of a side note... I've run a few comparisons and there are still some differences between OpenCL and R/Python, with OpenCL generally having more cases. I _think_ this might be down to some differences in the disease pathway. 

It seems as if individuals in the R/Python model stay in the exposed stage longer but those in the OpenCL stay in the symptomatic/asymptomatic stage longer. I guess this would mean that the OpenCL model has more individuals in an infectious state for longer and therefore is producing more hazard/cases.  Not sure why this would be as I think the timings are pretty much equivalent but I might try and dig into this a bit more. 

I'm sure there is some element of randomness in it but it does consistently seem to be the OpenCL case producing a higher number of cases.

**Comment by rory-improbable** [comment]
@spoonerf Hmm very interesting, I thought we had finally ironed out all the differences between the models! I suppose it would be worth double checking the duration distributions for those stages, perhaps there's some difference in the probability distributions, or we're using the parameters in different ways somehow. 

Actually now that I think about it the lognormal distribution for infection durations is parameterised differently from the R version, so I'm not sure if the parameter values turn out to be equivalent, I didn't do the calculations to make them exact, because I figured we might be playing around with the params of that lognormal distribution anyway.

If you run [this test function](https://github.com/Urban-Analytics/RAMP-UA/blob/c3101a375cd5ed5535f6168dc8aae0f5a6f976d5/tests/opencl/test_update_statuses_kernel.py#L505) with the `visualize` flag set to `True` then it will plot a histogram of the infection durations, so that might be a good place to start, eg. comparing that histogram to a histogram of the lognormal distribution in the R code

**Comment by spoonerf** [comment]
Hi @rory-improbable Yeh I have a feeling it's the lognormal one as well. Have had a quick look at the exposed and presymp and I can't see any real differences in how they are calculated/parameterised. 

Will check out that test function - cheers!

**Comment by spoonerf** [comment]
Ah yeh they are quite different:

Will have a think about how to make them more equivalent (maybe a job for monday!) - think the OpenCL shape is much better but too long-tailed. Would this be the kind of thing the automatic calibration could be useful for @nickmalleson ?

opencl_lognorm

rpython_lognormal

**Comment by rory-improbable** [comment]
(@spoonerf going to move this conversation to slack since it's slightly tangential to this PR)

**Comment by nickmalleson** [comment]
Thanks @spoonerf & @rory-improbable , I'll merge these now so that I can update the calibration notebooks. In answer to your question Fiona: yes I think this is exactly the kind of thing we could try to calibrate on. I'll message you on slack

**Pull request merged**